### PR TITLE
Load: modify TsFile version check to enable V3 load to V4

### PR DESCRIPTION
**Pull request by DanielWang2035**

### CODE REVIEW

**Review by luoluoyuyu — APPROVED** [review]

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/ModificationUtils.java [hunk 1]
```diff
@@ -194,9 +194,10 @@ public static boolean isPointDeleted(long timestamp, List<TimeRange> deletionLis
   public static boolean isDeviceDeletedByMods(
       Collection<Modification> modifications, IDeviceID device, long startTime, long endTime)
       throws IllegalPathException {
+    final MeasurementPath deviceWithWildcard = new MeasurementPath(device, ONE_LEVEL_PATH_WILDCARD);
     for (Modification modification : modifications) {
-      PartialPath path = modification.getPath();
-      if (path.include(new MeasurementPath(device, IoTDBConstant.ONE_LEVEL_PATH_WILDCARD))
+      MeasurementPath path = modification.getPath();
+      if (path.matchFullPath(deviceWithWildcard)
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/ModificationUtils.java:200]:
```suggestion
      if (deviceWithWildcard.matchFullPath(path)
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/ModificationUtils.java [hunk 2]
```diff
@@ -194,9 +194,10 @@ public static boolean isPointDeleted(long timestamp, List<TimeRange> deletionLis
   public static boolean isDeviceDeletedByMods(
       Collection<Modification> modifications, IDeviceID device, long startTime, long endTime)
       throws IllegalPathException {
+    final MeasurementPath deviceWithWildcard = new MeasurementPath(device, ONE_LEVEL_PATH_WILDCARD);
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/ModificationUtils.java:197]: Need further discussion here...

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/splitter/TsFileSplitter.java [hunk 3]
```diff
@@ -389,8 +389,20 @@ private boolean checkMagic(TsFileSequenceReader reader) throws IOException {
...
 
     byte versionNumber = reader.readVersionNumber();
-    if (versionNumber != TSFileConfig.VERSION_NUMBER) {
-      logger.error("the file's Version Number is incorrect, file path: {}", reader.getFileName());
+    if (versionNumber < TSFileConfig.VERSION_NUMBER) {
+      if (versionNumber < TSFileConfig.VERSION_NUMBER - 1) {
+        logger.error("the file's Version Number is too old, file path: {}", reader.getFileName());
+        return false;
+      } else {
+        logger.warn(
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/splitter/TsFileSplitter.java:397]:
```suggestion
        logger.info(
```

Comment (SteveYurongSu) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/splitter/TsFileSplitter.java:397]: Need to test loading v3 files into v4 IoTDB with type dismatch conditions.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/ModificationUtils.java (hunk 1 above)
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/utils/ModificationUtils.java:200]: I do not think calling matchFullPath here is correct after I read the Java doc.

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/LoadTsFileAnalyzer.java [hunk 4]
```diff
@@ -250,7 +255,16 @@ private void analyzeSingleTsFile(final File tsFile, final boolean isDeleteAfterL
         tsFileResource.deserialize();
       }
 
-      schemaAutoCreatorAndVerifier.setCurrentModificationsAndTimeIndex(tsFileResource);
+      final TsFileDeviceIterator tsFileDeviceIterator = reader.getAllDevicesIteratorWithIsAligned();
+      final AtomicInteger deviceCount = new AtomicInteger();
+      tsFileDeviceIterator.forEachRemaining(o -> deviceCount.getAndIncrement());
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/LoadTsFileAnalyzer.java:260]: Please reduce the cost here when there is no mods file

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/memory/LoadTsFileMemoryManager.java [hunk 5]
```diff
@@ -100,6 +100,31 @@ public synchronized LoadTsFileAnalyzeSchemaMemoryBlock allocateAnalyzeSchemaMemo
...
+      return;
+    }
+
+    long bytesNeeded = newSizeInBytes - memoryBlock.getTotalMemorySizeInBytes();
+    try {
+      forceAllocateFromQuery(bytesNeeded);
+    } catch (LoadRuntimeOutOfMemoryException e) {
+      if (dataCacheMemoryBlock != null && dataCacheMemoryBlock.doShrink(bytesNeeded)) {
+        memoryBlock.setTotalMemorySizeInBytes(newSizeInBytes);
+      }
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/memory/LoadTsFileMemoryManager.java:122]:
```suggestion
      if (dataCacheMemoryBlock != null && dataCacheMemoryBlock.doShrink(bytesNeeded)) {
        memoryBlock.setTotalMemorySizeInBytes(newSizeInBytes);
        return;
      }
```

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/memory/LoadTsFileMemoryManager.java [hunk 6]
```diff
@@ -100,6 +100,31 @@ public synchronized LoadTsFileAnalyzeSchemaMemoryBlock allocateAnalyzeSchemaMemo
...
+      throw e;
```
Comment (SteveYurongSu) [inline iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/memory/LoadTsFileMemoryManager.java:123]:
```suggestion
      if (dataCacheMemoryBlock == null || !dataCacheMemoryBlock.doShrink(bytesNeeded)) {
        throw e;
      }
```

**Comment by SteveYurongSu** [comment]
Plz bump the tsfile dependency to a newer version :_)

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/splitter/TsFileSplitter.java (hunk 3 above)
Comment (DanielWang2035) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/splitter/TsFileSplitter.java]: fixed

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/LoadTsFileAnalyzer.java (hunk 4 above)
Comment (DanielWang2035) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/queryengine/plan/analyze/LoadTsFileAnalyzer.java]: fixed

File: iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/memory/LoadTsFileMemoryManager.java (hunk 6 above)
Comment (DanielWang2035) [reply iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/load/memory/LoadTsFileMemoryManager.java]: fixed

**Review by SteveYurongSu — APPROVED** [review]

**Pull request merged**